File: solutions/java/src/main/java/com/epi/RotateArrayPermutation.java
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.
package com.epi;

public class RotateArrayPermutation {
  // @include
  public static void rotateArray(int rotateAmount, int[] A) {
    rotateAmount %= A.length;
    int numCycles = (int)GCD.elementaryGCD(A.length, rotateAmount);
    int cycleLength = A.length / numCycles;

    for (int c = 0; c < numCycles; ++c) {
      applyCyclicPermutation(rotateAmount, c, cycleLength, A);
    }
  }

  private static void applyCyclicPermutation(int rotateAmount, int offset,
                                             int cycleLength, int[] A) {
    int temp = A[offset];
    for (int i = 1; i < cycleLength; ++i) {
      int val = A[(offset + i * rotateAmount) % A.length];
      A[(offset + i * rotateAmount) % A.length] = temp;
      temp = val;
    }
    A[offset] = temp;
  }
  // @exclude
}

File: static/python/rotate_array_permutation.py
import fractions


# @include
def rotate_array(rotate_amount, A):
    def apply_cyclic_permutation(rotate_amount, offset):
        temp = A[offset]
        for i in range(1, cycle_length):
            idx = (offset + i * rotate_amount) % len(A)
            A[idx], temp = temp, A[idx]
        A[offset] = temp

    rotate_amount %= len(A)
    if rotate_amount == 0:
        return
    num_cycles = fractions.gcd(len(A), rotate_amount)
    cycle_length = len(A) // num_cycles
    for c in range(num_cycles):
        apply_cyclic_permutation(rotate_amount, c)
# @exclude
